### PR TITLE
Support stop testkit itself

### PR DESCRIPTION
**Pull request by jxnu-liguobin**
I'm not sure if I did it right
References #29780

### CODE REVIEW

File: akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala [hunk 1]
```diff
@@ -395,6 +399,19 @@ class BehaviorTestKitSpec extends AnyWordSpec with Matchers with LogCapturing {
...
+
+    "start and stop child and then stop testkit itself" in {
+      val testkit = BehaviorTestKit(Parent.init)
+      testkit.run(SpawnChild)
+      val child = testkit.expectEffectType[Spawned[String]]
+      testkit.run(StopChild(child.ref))
+      testkit.expectEffect(Stopped(child.childName))
+
+      val ok = testkit.stopSelf(StopSelf)
+      assertEquals(true, ok)
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala:411]: I'm not sure I understand the purpose of this feature. Couldn't this specific `StopSelf` message be tested with and ordinary `testkit.run(StopSelf)`?

The issue https://github.com/akka/akka/issues/29780 talks about the need to simulate that the parent stops a child actor, where the parent here would be the testkit itself. That would correspond to `context.stop(childRef)`. In that case I don't see why `def stopSelf(message)` should take a message parameter?

Comment (jxnu-liguobin) [reply akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala:411]: Thank you for your reply. In fact, I'm still considering it because I'm not familiar with akka, and I'll update PR later.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala [hunk 2]
```diff
@@ -172,6 +173,21 @@ private[akka] final class BehaviorTestKitImpl[T](
   override def clearLog(): Unit = context.clearLog()
 
   override def receptionistInbox(): TestInboxImpl[Receptionist.Command] = context.system.receptionistInbox
+
+  override def stopSelf: Boolean = {
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala:177]: I don't think there is a point in the `Boolean` here since it will always return `true` (!isAlive).
The signature could be
```
def stopSelf(): Unit
```

or simply
```
def stop(): Unit
```

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala [hunk 3]
```diff
@@ -172,6 +173,21 @@ private[akka] final class BehaviorTestKitImpl[T](
...
+    try {
+      context.setCurrentActorThread()
+      val actorRef = selfInbox().ref.asInstanceOf[ActorRef[AnyRef]]
+      actorRef.tell(Terminate) // is ok?
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala:181]: is it necessary to send this `Terminate`?

File: akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala [hunk 4]
```diff
@@ -395,6 +396,20 @@ class BehaviorTestKitSpec extends AnyWordSpec with Matchers with LogCapturing {
...
+      val testkit = BehaviorTestKit(Parent.init)
+      testkit.run(SpawnChild)
+      val child = testkit.expectEffectType[Spawned[String]]
+      testkit.run(StopChild(child.ref))
+      testkit.expectEffect(Stopped(child.childName))
+
+      // stop testkit
+      val check = testkit.stopSelf
+      // testkit has no effect
+      assertEquals(true, check)
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala:410]: the verification here could be
```
testkit.isAlive shouldBe false
```

File: akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala [hunk 5]
```diff
@@ -395,6 +396,20 @@ class BehaviorTestKitSpec extends AnyWordSpec with Matchers with LogCapturing {
       val newChild = testkit.expectEffectType[Spawned[_]]
       child.childName shouldBe newChild.childName
     }
+
+    "start and stop child and then stop testkit itself" in {
+      val testkit = BehaviorTestKit(Parent.init)
+      testkit.run(SpawnChild)
+      val child = testkit.expectEffectType[Spawned[String]]
+      testkit.run(StopChild(child.ref))
+      testkit.expectEffect(Stopped(child.childName))
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala:405]: This part of the test is exactly the same as in "stop and restart a named child". Do we have to repeat that test here? How is that related to the `stopSelf`?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKit.scala [hunk 6]
```diff
@@ -155,4 +155,9 @@ trait BehaviorTestKit[T] {
    * The receptionist inbox contains messages sent to `system.receptionist`
    */
   def receptionistInbox(): TestInbox[Receptionist.Command]
+
+  /**
+   * Stop this testkit and return true if this testkit stop successfully.
+   */
+  def stopSelf: Boolean
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKit.scala:162]: should also add corresponding in `javadsl.BehaviorTestKit`

**Comment by patriknw** [comment]
@leviramsey could you review this and suggest how to move it forward?

**Comment by jxnu-liguobin** [comment]
> @leviramsey could you review this and suggest how to move it forward?

yes, I will update it.

File: akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala (hunk 5 above)
Comment (jxnu-liguobin) [reply akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKitSpec.scala:405]: The difference is the hasEffects, the new pr has added the stop effect of the current actor in the stopSelf method.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala [hunk 7]
```diff
@@ -172,6 +173,18 @@ private[akka] final class BehaviorTestKitImpl[T](
   override def clearLog(): Unit = context.clearLog()
 
   override def receptionistInbox(): TestInboxImpl[Receptionist.Command] = context.system.receptionistInbox
+
+  override def stopSelf(): Unit = {
+    try {
+      context.setCurrentActorThread()
+      currentUncanonical = Behaviors.stopped
+      current = Behaviors.stopped
+      context.effectQueue.clear()
```
Comment (leviramsey) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala:182]: I'm not sure the effect queue should be cleared here.  I don't see the reason to prohibit something like:

```scala
// run command and then stop
testkit.run(someCommand)
testkit.stopSelf()

textkit.expectEffectType[Effect.Stopped]
```

I could definitely see how a "Given-When-Then" test might be structured that way ("When I run this command and then stop").

Comment (leviramsey) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala:182]: The same applies to `clearLog`.
